N°3422 - Show attachments metadata when choosing the one to send by email

mraenker opened the pull request:
Hi Combodo!

### Summary ###
This PR adds more info about the file one is about to choose to attach to a notification-mail. Namely the uploaders name (if given), the time, when it was uploaded (if given), and the filesize.

### Screenshots ###
- This is how it renders in an "optimal" case (all info given, no extraordinary long names and so on):
![image](https://user-images.githubusercontent.com/21227916/213553460-11d819cd-1018-47f6-9216-b9a088221fa6.png)

- This is how it renders in a case, where the filenames are too long/the viewport is too small:
![image](https://user-images.githubusercontent.com/21227916/213553560-3d60a6b0-d79c-4e9a-ad82-e801a3232252.png)

- This is how it renders, if the uploaders name isn't given (note the autoshrinking and no unncessary stuff is rendered):
![image](https://user-images.githubusercontent.com/21227916/213553841-020bd28b-9828-4802-a690-f410b7999da2.png)

- This is how it renders, if the timestamp isn't given (note the autoshrinking and no unncessary stuff is rendered):
![image](https://user-images.githubusercontent.com/21227916/213553919-05922def-4261-4870-b901-9cc854ba2eef.png)

### Notes ###
- Tested on Firefox and Chrome.
- There a few cases, where the timestamp or the person on an attachment aren't recorded correctly. In this case, they are not displayed.
- The resulting filenamestring can be rather long, that's why I opted to fatten the filename itself.
- This is the reason for using `utils::HtmlEntities()` on the filename itself and on the persons name (omitting timestamp and filesize, since they are calculated serverside and can't be edited later on) and then removing `htmlentitites()` in line 105 of ajax.php for the filename. I don't know, if it's good practice to escape the timestamp and filesize aswell. Feel free to add it, if needed.
- Since I want the strings to be as short as possible, I opted to not use the Dicts `Attachments:File:Date` and `Attachments:File:Uploader`, since they are rather long in some languages, but instead created new entries within the Extension for the little words `by` and `at` (and I hope, there are short equivalents in nearly all languages).
- When creating the modal, I changed the definitions for `width`, `maxWidth` and `minWidth` to be all relative now. The behavior is (or better: should be; I am no expert on CSS whatsoever): (a) Make the modal as small as possible without linebreaks, (b) but only grow it to 90% of the viewports(?) width and (c) make it not indefinitely small.

Martin

Molkobain commented on ajax.php:
```diff
@@ -76,11 +76,21 @@
+				$sNameUploader = utils::HtmlEntities($oAttachment->Get('contact_id_friendlyname'));
+				$sTimeStampUpload = $oAttachment->Get('creation_date');
+				$sFileFormattedSize = $oDoc->GetFormattedSize();
+				$sFilename = utils::HtmlEntities($sFileName);
+				$sFileName = '<b>'.$sFileName.'</b>';
+				$sFileName .= $sNameUploader != '' ? ' '.Dict::S('UI-emry-attachment-by').' '.$sNameUploader : '';
+				$sFileName .= $sTimeStampUpload != '' ? ' '.Dict::S('UI-emry-attachment-at').' '.$sTimeStampUpload : '';
+				$sFileName .= ' - '.$sFileFormattedSize;
```
Mind to use `utils::HtmlEntities()` for the uploader name as well as it could contain special chars.

_Note: Since iTop 3.0 we rather use `utils::EscapeHtml()`, but as this extension must remain compatible with iTop 2.7 we can't use it here._

Also, I would feel more comfortable if we rename `$sFilename` as it is something more now. "$sAttachementLabel" or something like that.

Finally, we can't insert the "at" and "by" like that as in some language it might be a totally different arrangement.
The best approach (to me) would be to have different dict entries for the differents combinaisons. It might be more redundant but also more flexible for all languages.

mraenker replied:
Thank you, will be done. A question about this part:

> Finally, we can't insert the "at" and "by" like that as in some language it might be a totally different arrangement. The best approach (to me) would be to have different dict entries for the differents combinaisons. It might be more redundant but also more flexible for all languages.

If that needs to happen: Do you think, it's ok to move all the code for "making" the label into an own function in ajax.php somewhere above, as I think, there will be quite a few more ifs and elses and the block of code we are talking about already is a bit big?

Hipska replied:
Yeah, it would be nice if there would be a method that was given the attachment and gives the correct text in return..

Molkobain replied:
It can be in a method yes, but it must use dict entries no matter what so it can arranged correctly in any language.
Here is what I had in mind:
```
$sFilename = utils::HtmlEntities($sFileName);
$bHasUploaderName = $sNameUploader != '';
$bHasUploadTimestamp = $sTimeStampUpload != '';
if ($bHasUploaderName && $bHasUploadTimestamp) {
	$sDictEntryCode = 'A';
} elseif ($bHasUploaderName && !$bHasUploadTimestamp) {
	$sDictEntryCode = 'B';
} elseif (!$bHasUploaderName && $bHasUploadTimestamp) {
	$sDictEntryCode = 'C';
} else {
	$sDictEntryCode = 'D';
}
$sAttachementLabel = Dict::Format($sDictEntryCode, $sFilename, $sNameUploader, $sTimeStampUpload, $sFileFormattedSize);
```

With dict entries being like
```
'A' => '<b>%1$s</b> by %2$s at %3$s - %4$s',
'B' => '<b>%1$s</b> by %2$s - %4$s',
'C' => '<b>%1$s</b> at %3$s - %4$s',
'D' => '<b>%1$s</b> - %4$s',
```

_Edit: It's not great to have markup in the dict entries, but we already have some entries like this._

Hipska replied:
To reduce the number of variants, I think only the filename and date, would be enough?

Molkobain replied:
Can you give an example of that you mean? I'm not sure to understand.

Hipska replied:
![](https://user-images.githubusercontent.com/21227916/213553841-020bd28b-9828-4802-a690-f410b7999da2.png)

Molkobain replied:
It's up to Martin really, it's fine by me in both cases.
Any thoughts @v-dumas ?

Molkobain reviewed (changes requested):
For for the PR Martin!

piRGoif commented:
Hello,
Very good suggestion, thanks !
Did you try to display in an array, maybe in a wider modal ?

mraenker commented:
> Hello, Very good suggestion, thanks ! Did you try to display in an array, maybe in a wider modal ?

I don't quite understand, what you mean by "array", I think. Do you mean an HTML Table? What I can say: The wider the modal, the more "ugly" the user experience becomes (even more so with low resolutions or on mobile); that's why I tried to keep the lines as short as possible.

piRGoif commented:
Yes, I meant HTML table, the same one displayed in tickets.
I imagine they could fit well in a smaller display space ? 
Unfortunately I don't think we have yet a very generic code to generate this table, so might be hard work :/ But giving a try with static html might be worse a try ?

mraenker commented:
> Yes, I meant HTML table, the same one displayed in tickets. I imagine they could fit well in a smaller display space ? Unfortunately I don't think we have yet a very generic code to generate this table, so might be hard work :/ But giving a try with static html might be worse a try ?

Yep. We discussed this with Guillaume and Hipska on Slack beforehand and agreed to first make this "simple" PR happen and later on do the "hard work", where we implement a proper renderer with all the JS and stuff encapsuled at the "right" place, because tha latter I won't be able to provide in the next few weeks.

(See https://app.slack.com/client/T02CP4UD56D/C02CAG5MLH0/thread/C02CAG5MLH0-1673948720.871579)

piRGoif commented:
Oh sorry I missed the discussion on Slack.
Fine by me, ok !

mraenker commented:
- `utils::HtmlEntitites()` for Uploaders Name and original Filename
- Renamed the var from `$sFileName` to `$sAttachmentLabel`
- Moved the functionality into its own function `MakeAttachmentLabel()` (which takes the `ormDocument` and the `iObjId` (not the attachment itself) as arguments, because that's what we have in the main part of the code and i wanted to not clutter the main part with getting the attachment via `MetaModel::GetObject()`
- Used @Molkobain suggestion with the different cases (all tested), but slightly modified: the possibility of both uploaders name and timestamp ~was missing~ was encoded in adding the ~timestamp~ filesize, but we can omit the case with the filesize, since it will be always available and we can append it at the end (no need to include it in the dict aswell). So we still have 4 cases, but the dicts have one argument less.
- en and de Dicts are updated

@Hipska :
> To reduce the number of variants, I think only the filename and date, would be enough?

I thought about this longer, than I expected, because I think, you are technically right: The upload date is a nice identifier, which should be enough on its own. But I think, what humans actually do is to (subconsciously) use every bit of info they can get to make a faster decision. If I imagine, after this change, I _still_ would have to parse the date first and check it against the upload-list in another tab, while it actually _would be_ possible, to simply look at different names or filesizes in the same modal, if they were presented, I would be a bit annoyed. So I vote for "present the user with everything, which helps to make an easier decision". Do you agree?

Have a nice weekend.

Martin

Hipska commented:
Yes I agree, It was only a suggestion to have fewer dictionary items for this (temporary) feature until the big feature with the nice table comes around..

Hipska commented on ajax.php:
```diff
@@ -79,8 +113,10 @@
 				$oDoc = $oObj->Get($aData['sBlobAttCode']);
-				$sFileName = $oDoc->GetFileName();
-				$sIcon = utils::GetAbsoluteUrlAppRoot().AttachmentPlugIn::GetFileIcon($sFileName);
+				// The unmodified filename is needed first to choose the icon right after, then we make right label
+				$sAttachmentLabel = $oDoc->GetFileName();
+				$sIcon = utils::GetAbsoluteUrlAppRoot().AttachmentPlugIn::GetFileIcon($sAttachmentLabel);
```
I would revert this part as the `$sAttachmentLabel` gets overwritten anyway.
```suggestion
				$sFileName = $oDoc->GetFileName();
				$sIcon = utils::GetAbsoluteUrlAppRoot().AttachmentPlugIn::GetFileIcon($sFileName);
```

Hipska commented on ajax.php:
```diff
@@ -16,6 +16,40 @@
 
+function MakeAttachmentLabel($oOrmDoc, $iObjId)
+{
+	$oAttachment = MetaModel::GetObject('Attachment', $iObjId);
```
Hardcoded classname? The code around the method invocation has `$sObjClass` available, so I would pass and use that.

Or actually pass the `$oAttachment` directly.

Hipska replied:
You did mark as resolved, but didn't see any change or comment/response on this?

Hipska commented on en.dict.email-reply.php:
```diff
@@ -39,5 +39,9 @@
 	'UI-emry-select-attachments:Short' => 'Attachments',
 	'UI-emry-attachments-to-be-sent' => 'The following attachments will be sent:',
 	'UI-emry-select-attachments-tooltip' => 'Click to select the attachments to be sent',
+	'UI-emry-attachment-label-case-A' => '<b>%1$s</b> by %2$s at %3$s',
+	'UI-emry-attachment-label-case-B' => '<b>%1$s</b> by %2$s',
+	'UI-emry-attachment-label-case-C' => '<b>%1$s</b> at %3$s',
+	'UI-emry-attachment-label-case-D' => '<b>%1$s</b>',
```
I also think the naming `case-A` to `case-D` were just examples, not to be actually used?

Molkobain replied:
```suggestion
	'UI-emry-attachment-label-with-uploadername-and-timestamp' => '<b>%1$s</b> by %2$s at %3$s',
	'UI-emry-attachment-label-with-uploadername' => '<b>%1$s</b> by %2$s',
	'UI-emry-attachment-label-with-timestamp' => '<b>%1$s</b> at %3$s',
	'UI-emry-attachment-label' => '<b>%1$s</b>',
```

Molkobain commented on ajax.php:
```diff
@@ -16,6 +16,40 @@
+	$bHasUploadTimestamp = $sTimeStampUpload != '';
+	if ($bHasUploaderName && $bHasUploadTimestamp)
+	{
+		$sDictEntryCode = 'UI-emry-attachment-label-case-A';
```
```suggestion
		$sDictEntryCode = 'UI-emry-attachment-label-with-uploadername-and-timestamp';
```

Hipska replied:
If you would name it `-with-uploadername-with-timestamp`, it could even be programatically easier to select the correct dict. (1 if statement less)

Molkobain replied:
But it would feel weird don't you think? Not sure it's worth it only remove 1 "if" statement.

Hipska replied:
This looks much more clean, isn't it?

```php
	$sDictEntryCode = 'UI-emry-attachment-label';
	if ($bHasUploaderName) $sDictEntryCode .= '-with-uploadername';
	if ($bHasUploadTimestamp) $sDictEntryCode .= '-with-timestamp';
```

Molkobain replied:
Why not, but no one-line-if-condition (see conventions) :)

Hipska replied:
Hmm, but this is allowed in the conventions:

```php
	$sDictEntryCode  = 'UI-emry-attachment-label';
	$sDictEntryCode .= ($bHasUploaderName) ? '-with-uploadername' : '';
	$sDictEntryCode .= ($bHasUploadTimestamp) ? '-with-timestamp' : '';
```

Molkobain replied:
Yes.

Molkobain commented on ajax.php:
```diff
@@ -16,6 +16,40 @@
+	}
+	elseif ($bHasUploaderName && !$bHasUploadTimestamp)
+	{
+		$sDictEntryCode = 'UI-emry-attachment-label-case-B';
```
```suggestion
		$sDictEntryCode = 'UI-emry-attachment-label-with-uploadername';
```

Molkobain commented on ajax.php:
```diff
@@ -16,6 +16,40 @@
+	}
+	elseif (!$bHasUploaderName && $bHasUploadTimestamp) 
+	{
+		$sDictEntryCode = 'UI-emry-attachment-label-case-C';
```
```suggestion
		$sDictEntryCode = 'UI-emry-attachment-label-with-timestamp';
```

Molkobain commented on ajax.php:
```diff
@@ -16,6 +16,40 @@
+	}
+	else
+	{
+		$sDictEntryCode = 'UI-emry-attachment-label-case-D';
```
```suggestion
		$sDictEntryCode = 'UI-emry-attachment-label';
```

Molkobain commented on de.dict.email-reply.php:
```diff
@@ -38,5 +38,9 @@
 	'UI-emry-select-attachments:Short' => 'Anhänge...',
 	'UI-emry-attachments-to-be-sent' => 'Die folgenden Anhänge werden versandt:',
 	'UI-emry-select-attachments-tooltip' => 'Klicken Sie, um die zu sendenden Anhänge auszuwählen.',
+	'UI-emry-attachment-label-case-A' => '<b>%1$s</b> von %2$s vom %3$s',
+	'UI-emry-attachment-label-case-B' => '<b>%1$s</b> von %2$s',
+	'UI-emry-attachment-label-case-C' => '<b>%1$s</b> vom %3$s',
+	'UI-emry-attachment-label-case-D' => '<b>%1$s</b>',
```
```suggestion
	'UI-emry-attachment-label-with-uploadername-and-timestamp' => '<b>%1$s</b> von %2$s vom %3$s',
	'UI-emry-attachment-label-with-uploadername' => '<b>%1$s</b> von %2$s',
	'UI-emry-attachment-label-with-timestamp' => '<b>%1$s</b> vom %3$s',
	'UI-emry-attachment-label' => '<b>%1$s</b>',
```

Molkobain reviewed (changes requested):
Thanks for all the changes. Just one thing, when I said "A, B, C, D" in the cases, it was a lazy example. We need to keep dict entries understandable in their codes.

mraenker committed:
N°3422 - Add Uploader, Timestamp and Filesize to Filechooser-Modal

mraenker commented:
All done, rebased and pushed again.

Molkobain commented:
Great thanks! Do you feel like it would be good to have the formatted size as a placeholder in the dict entries as well or is it overkill?

mraenker commented:
> Great thanks! Do you feel like it would be good to have the formatted size as a placeholder in the dict entries as well or is it overkill?

- I intentionally "voted" against it (in my head). Reason: if we got no info (as in: no uploaders name and no timestamp), the filesize (which will always be available), will be the "last resort" of additional information.
- `KB` and the likes are internationally used abbreviations, which (afaik) are not translated, at least in Europe. (Do e. g. Chinese or Japanese people use their own symbols for that? Don't know, but I think, even if, it would be overkill.) 

But I think, that's ultimately up to you.

Hipska commented:
You forgot of course the French where they use O for Octets 😂

mraenker commented:
> You forgot of course the French where they use O for Octets

Wow. TIL.

Well, then I think, it's even more up to the folks from Combodo to decide. ;) :D

Molkobain commented:
Actually I was thinking about something more like if someone wants to move the file size elsewhere in the label (eg. next to the filename  `some-picture.jpg (132kB) - from Martin on 2023-01-01`).
If it is in the dict entry like everything else, it will be very easy for any customer / community user. :)

We would definitively prefer it this way, sorry for "another change"

_Edit: Just add all suggestions so you can apply the changes easily_

Molkobain commented on ajax.php:
```diff
@@ -16,6 +16,35 @@
+	{
+		$sDictEntryCode .= '-with-timestamp';
+	}
+	$sAttachmentLabel = Dict::Format($sDictEntryCode, $sAttachmentLabel, $sNameUploader, $sTimeStampUpload);
+	$sAttachmentLabel .= ' - '.$sFileFormattedSize;
```
```suggestion
	$sAttachmentLabel = Dict::Format($sDictEntryCode, $sFilenameForHtml, $sFileFormattedSize, $sNameUploader, $sTimeStampUpload);
```

Molkobain commented on ajax.php:
```diff
@@ -16,6 +16,35 @@
+	$sTimeStampUpload = $oAttachment->Get('creation_date');
+	$sFileFormattedSize = $oOrmDoc->GetFormattedSize();
+	$sNameUploader = utils::HtmlEntities($oAttachment->Get('contact_id_friendlyname'));
+	$sAttachmentLabel = $oOrmDoc->GetFileName();
+	$sAttachmentLabel = utils::HtmlEntities($sAttachmentLabel);
```
```suggestion
	$sFilename = $oOrmDoc->GetFileName();
	$sFilenameForHtml = utils::HtmlEntities($sFilename);
```

Molkobain commented on de.dict.email-reply.php:
```diff
@@ -38,5 +38,9 @@
 	'UI-emry-select-attachments:Short' => 'Anhänge...',
 	'UI-emry-attachments-to-be-sent' => 'Die folgenden Anhänge werden versandt:',
 	'UI-emry-select-attachments-tooltip' => 'Klicken Sie, um die zu sendenden Anhänge auszuwählen.',
+	'UI-emry-attachment-label-with-uploadername-and-timestamp' => '<b>%1$s</b> von %2$s vom %3$s',
+	'UI-emry-attachment-label-with-uploadername' => '<b>%1$s</b> von %2$s',
+	'UI-emry-attachment-label-with-timestamp' => '<b>%1$s</b> vom %3$s',
+	'UI-emry-attachment-label' => '<b>%1$s</b>',
```
```suggestion
	'UI-emry-attachment-label-with-uploadername-and-timestamp' => '<b>%1$s</b> von %3$s vom %4$s - %2$s',
	'UI-emry-attachment-label-with-uploadername' => '<b>%1$s</b> von %3$s - %2$s',
	'UI-emry-attachment-label-with-timestamp' => '<b>%1$s</b> vom %4$s - %2$s',
	'UI-emry-attachment-label' => '<b>%1$s</b> - %2$s',
```

Molkobain commented on en.dict.email-reply.php:
```diff
@@ -39,5 +39,9 @@
 	'UI-emry-select-attachments:Short' => 'Attachments',
 	'UI-emry-attachments-to-be-sent' => 'The following attachments will be sent:',
 	'UI-emry-select-attachments-tooltip' => 'Click to select the attachments to be sent',
+	'UI-emry-attachment-label-with-uploadername-and-timestamp' => '<b>%1$s</b> by %2$s at %3$s',
+	'UI-emry-attachment-label-with-uploadername' => '<b>%1$s</b> by %2$s',
+	'UI-emry-attachment-label-with-timestamp' => '<b>%1$s</b> at %3$s',
+	'UI-emry-attachment-label' => '<b>%1$s</b>',
```
```suggestion
	'UI-emry-attachment-label-with-uploadername-and-timestamp' => '<b>%1$s</b> by %3$s at %4$s - %2$s',
	'UI-emry-attachment-label-with-uploadername' => '<b>%1$s</b> by %3$s - %2$s',
	'UI-emry-attachment-label-with-timestamp' => '<b>%1$s</b> at %4$s - %2$s',
	'UI-emry-attachment-label' => '<b>%1$s</b> - %2$s',
```

Molkobain commented:
> You forgot of course the French where they use O for Octets 😂

True 😅😅
And I just saw that the `utils::BytesToFriendlyFormat` function doesn't localize the units... But that's another battle 😁

mraenker committed:
Update ajax.php

Co-authored-by: Molkobain <lajarige.guillaume@free.fr>

mraenker committed:
Update ajax.php

Co-authored-by: Molkobain <lajarige.guillaume@free.fr>

mraenker committed:
Update de.dict.email-reply.php

Co-authored-by: Molkobain <lajarige.guillaume@free.fr>

mraenker committed:
Update en.dict.email-reply.php

Co-authored-by: Molkobain <lajarige.guillaume@free.fr>

mraenker commented:
> Edit: Just add all suggestions so you can apply the changes easily

Thank you! Done.

Molkobain commented:
Thanks for all the changes! 🙏🙌

piRGoif commented:
Hello,
About dict entries parameters, handling RTL would be a reason to keep parameters, so that order can be changed. For now we don't have such language in iTop, so maybe we can skip this specific need ?

piRGoif commented:
About Bytes VS octets etc, we are using massively in iTop \utils::BytesToFriendlyFormat, which doesn't handle translations : it always returns the english name...
And it's already the one used in attachments (TableDetailsAttachmentsRenderer calling \ormDocument::GetFormattedSize, calling \utils::BytesToFriendlyFormat)

Molkobain commented:
Accepted during technical review.

Molkobain commented:
Accepted during functional review.
_(But will be merged only once iTop 3.0.3 is released)_

Hipska commented on ajax.php:
```diff
@@ -16,6 +16,34 @@
+	elseif (!$bHasUploaderName && $bHasUploadTimestamp) 
+	{
+		$sDictEntryCode .= '-with-timestamp';
+	}
```
```suggestion
	$sDictEntryCode .= ($bHasUploaderName) ? '-with-uploadername' : '';
	$sDictEntryCode .= ($bHasUploadTimestamp) ? '-with-timestamp' : '';
```

Hipska commented on de.dict.email-reply.php:
```diff
@@ -38,5 +38,9 @@
 	'UI-emry-select-attachments:Short' => 'Anhänge...',
 	'UI-emry-attachments-to-be-sent' => 'Die folgenden Anhänge werden versandt:',
 	'UI-emry-select-attachments-tooltip' => 'Klicken Sie, um die zu sendenden Anhänge auszuwählen.',
+	'UI-emry-attachment-label-with-uploadername-and-timestamp' => '<b>%1$s</b> von %3$s vom %4$s - %2$s',
+	'UI-emry-attachment-label-with-uploadername' => '<b>%1$s</b> von %3$s - %2$s',
+	'UI-emry-attachment-label-with-timestamp' => '<b>%1$s</b> vom %4$s - %2$s',
+	'UI-emry-attachment-label' => '<b>%1$s</b> - %2$s',
```
```suggestion
	'UI-emry-attachment-label-with-uploadername-with-timestamp' => '<b>%1$s</b> von %3$s vom %4$s - %2$s',
	'UI-emry-attachment-label-with-uploadername' => '<b>%1$s</b> von %3$s - %2$s',
	'UI-emry-attachment-label-with-timestamp' => '<b>%1$s</b> vom %4$s - %2$s',
	'UI-emry-attachment-label' => '<b>%1$s</b> - %2$s',
```

Hipska commented on en.dict.email-reply.php:
```diff
@@ -39,5 +39,9 @@
 	'UI-emry-select-attachments:Short' => 'Attachments',
 	'UI-emry-attachments-to-be-sent' => 'The following attachments will be sent:',
 	'UI-emry-select-attachments-tooltip' => 'Click to select the attachments to be sent',
+	'UI-emry-attachment-label-with-uploadername-and-timestamp' => '<b>%1$s</b> by %3$s at %4$s - %2$s',
+	'UI-emry-attachment-label-with-uploadername' => '<b>%1$s</b> by %3$s - %2$s',
+	'UI-emry-attachment-label-with-timestamp' => '<b>%1$s</b> at %4$s - %2$s',
+	'UI-emry-attachment-label' => '<b>%1$s</b> - %2$s',
```
```suggestion
	'UI-emry-attachment-label-with-uploadername-with-timestamp' => '<b>%1$s</b> by %3$s at %4$s - %2$s',
	'UI-emry-attachment-label-with-uploadername' => '<b>%1$s</b> by %3$s - %2$s',
	'UI-emry-attachment-label-with-timestamp' => '<b>%1$s</b> at %4$s - %2$s',
	'UI-emry-attachment-label' => '<b>%1$s</b> - %2$s',
```

Hipska reviewed (changes requested):
See https://github.com/Combodo/email-reply/pull/1#discussion_r1130982320

Molkobain commented:
@mraenker I'm about to merge this PR. Do you wish to apply / test Hipska's suggestions before?

mraenker committed:
Update ajax.php

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@me.com>

mraenker committed:
Update de.dict.email-reply.php

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@me.com>

mraenker committed:
Update en.dict.email-reply.php

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@me.com>

mraenker commented:
@Molkobain Done!

Hipska commented on ajax.php:
```diff
@@ -16,6 +16,24 @@
+	$sNameUploader = utils::HtmlEntities($oAttachment->Get('contact_id_friendlyname'));
+	$sFilename = $oOrmDoc->GetFileName();
+	$sFilenameForHtml = utils::HtmlEntities($sFilename);
+	$bHasUploaderName = $sNameUploader != '';
+	$bHasUploadTimestamp = $sTimeStampUpload != '';
+	$sDictEntryCode = 'UI-emry-attachment-label';
+	$sDictEntryCode .= ($bHasUploaderName) ? '-with-uploadername' : '';
+	$sDictEntryCode .= ($bHasUploadTimestamp) ? '-with-timestamp' : '';
```
I think this is now also clear enough, the variables aren't used for anything else..
```suggestion
	$sDictEntryCode = 'UI-emry-attachment-label';
	$sDictEntryCode .= ($sNameUploader != '') ? '-with-uploadername' : '';
	$sDictEntryCode .= ($sTimeStampUpload != '') ? '-with-timestamp' : '';
```

Hipska commented on ajax.php:
```diff
@@ -81,6 +99,7 @@
 				$oDoc = $oObj->Get($aData['sBlobAttCode']);
 				$sFileName = $oDoc->GetFileName();
 				$sIcon = utils::GetAbsoluteUrlAppRoot().AttachmentPlugIn::GetFileIcon($sFileName);
+				$sAttachmentLabel = MakeAttachmentLabel($oDoc, $sObjClass, $iObjId);
```
Since the `Attachment` object is available why not do this?

```suggestion
				$sAttachmentLabel = MakeAttachmentLabel($oDoc, $oObj);
```
